src/storage/src/storage/bidi/connector.rs

```diff
@@ -0,0 +1,842 @@
+        options: &RequestOptions,
+        api_client_header: &'static str,
+        request_params: &str,
+    ) -> impl Future<Output = Result<tonic::Result<tonic::Response<Self::Stream>>>> + Send;
+}
+
+impl Client for GrpcClient {
+    type Stream = tonic::codec::Streaming<BidiReadObjectResponse>;
+    async fn start(
+        &self,
+        extensions: tonic::Extensions,
+        path: http::uri::PathAndQuery,
+        rx: Receiver<BidiReadObjectRequest>,
+        options: &RequestOptions,
+        api_client_header: &'static str,
+        request_params: &str,
+    ) -> Result<tonic::Result<tonic::Response<Self::Stream>>> {
+        let request = tokio_stream::wrappers::ReceiverStream::new(rx);
+        self.bidi_stream_with_status(
+            extensions,
+            path,
+            request,
+            options.gax(),
+            api_client_header,
+            request_params,
+        )
+        .await
+    }
+}
+
+pub trait TonicStreaming: std::fmt::Debug + Send + 'static {
+    async fn next_message(&mut self) -> tonic::Result<Option<BidiReadObjectResponse>>;
+}
+
+impl TonicStreaming for tonic::codec::Streaming<BidiReadObjectResponse> {
+    async fn next_message(&mut self) -> tonic::Result<Option<BidiReadObjectResponse>> {
+        self.message().await
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use super::super::tests::{permanent_error, redirect_handle, redirect_status, test_options};
+    use super::*;
+    use crate::google::storage::v2::{BidiReadHandle, Object, ObjectRangeData};
+    use crate::read_resume_policy::{AlwaysResume, ReadResumePolicyExt};
+    use anyhow::Result;
+    use gax::error::binding::{BindingError, SubstitutionFail};
+    use gax::retry_policy::NeverRetry;
+    use static_assertions::assert_impl_all;
+    use std::error::Error as _;
+    use std::sync::Arc;
+    use test_case::test_case;
+
+    #[test]
+    fn assertions() {
+        assert_impl_all!(Connector: Clone, std::fmt::Debug, Send, Sync);
+    }
+
+    #[tokio::test]
+    async fn bad_endpoint() -> Result<()> {
+        fn need_send<T: Send>(_val: &T) {}
+
+        let mut config = gaxi::options::ClientConfig::default();
+        config.cred = Some(test_credentials());
+        let client = GrpcClient::new(config, "http://127.0.0.1:1").await?;
+
+        let spec = BidiReadObjectSpec {
+            bucket: "projects/_/buckets/test-only-bucket".into(),
+            object: "test-only-object".into(),
+            ..BidiReadObjectSpec::default()
+        };
+
+        let mut options = test_options();
+        options.retry_policy = Arc::new(NeverRetry);
+        let mut connector = Connector::new(spec, options, client);
+        let start = connector.connect(Vec::new());
+        need_send(&start);
+
+        let err = start.await.unwrap_err();
+        assert!(err.is_connect(), "{err:?}");
+
+        Ok(())
+    }
+
+    #[tokio::test]
+    #[test_case("")]
+    #[test_case("my-bucket")]
+    async fn binding(bucket_name: &str) -> Result<()> {
```

Comment:
nit: binding_error is slightly more informative of the set of tests. Was initially confused as to why expect_start was never.